### PR TITLE
typegen

### PR DESCRIPTION
**Pull request by forgetso**
- generate definitions from contract ABI
- fix error handling and add definitions into types key in output
- Change definitionGen to output u64 for timestamps and Map for Mapping
- Add common repository
- Add auto generated interfaces from polkadot typegen
- tests
- get frontend working

### CODE REVIEW

File: src/contract/helpers.ts [hunk 1]
```diff
@@ -45,54 +48,38 @@ export function getEventsFromMethodName(
...
     const typesToHash = ['Hash']
     methodObj.args.forEach((methodArg, idx) => {
-        const argVal = args[idx]
+        let argVal = args[idx]
         // hash values that have been passed as strings
         if (typesToHash.indexOf(methodArg.type.type) > -1 && !(isU8a(argVal) || isHex(argVal))) {
-            encodedArgs.push(stringToHexPadded(argVal as unknown as string) as unknown as T)
-        } else {
-            encodedArgs.push(argVal)
+            argVal = stringToHexPadded(argVal)
```
Comment (goastler) [inline src/contract/helpers.ts:62]: do we need a string length check? As the hashing function can only support strings of max length X

File: src/interfaces/definitions.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+export const definitions = {
```
Comment (goastler) [inline src/interfaces/definitions.ts:1]: is this generated from the ABI? If so, where?

**Review by goastler — CHANGES_REQUESTED** [review]

File: src/interfaces/definitions.ts (hunk 2 above)
Comment (goastler) [reply src/interfaces/definitions.ts:1]: nvm, found where it is generated. Are the definitions regenerated during build?

Comment (forgetso) [reply src/interfaces/definitions.ts:1]: The definitions are generated during runtime and added to the api. This file was only used to create the interface types which were then committed to the contracts repo.

File: src/contract/helpers.ts (hunk 1 above)
Comment (forgetso) [reply src/contract/helpers.ts:62]: The check is inside the function

```

    if (data.length > maxLength) {
        throw new Error(`stringToHexPadded: string length ${data.length} exceeds ${maxLength}`)
    }

```

Do you think I should catch this inside helpers.ts?

File: src/interfaces/definitions.ts (hunk 2 above)
Comment (goastler) [reply src/interfaces/definitions.ts:1]: Ah okay I see

File: src/contract/helpers.ts (hunk 1 above)
Comment (goastler) [reply src/contract/helpers.ts:62]: No I think it's good how it is. The error should be caught by the calling function that provided the string in the first place, which encodeStringArgs is not (just gets passed the string(s) as parameters)

**Review by goastler — APPROVED** [review]

**Pull request merged**